src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpOnTypeFormattingPass.cs

```diff
@@ -146,9 +149,54 @@ public async override Task<FormattingResult> ExecuteAsync(FormattingContext cont
             var finalChanges = cleanedText.GetTextChanges(originalText);
             var finalEdits = finalChanges.Select(f => f.AsTextEdit(originalText)).ToArray();
 
+            if (context.AutomaticallyAddUsings)
+            {
+                // Because we need to parse the C# code twice for this operation, lets do a quick check to see if its even necessary
+                if (result.Edits.Any(e => e.NewText.IndexOf("using") != -1))
+                {
+                    finalEdits = await AddUsingStatementEditsAsync(codeDocument, finalEdits, csharpText, originalTextWithChanges, cancellationToken);
+                }
+            }
+
             return new FormattingResult(finalEdits);
         }
 
+        private async Task<TextEdit[]> AddUsingStatementEditsAsync(RazorCodeDocument codeDocument, TextEdit[] finalEdits, SourceText csharpText, SourceText originalTextWithChanges, CancellationToken cancellationToken)
+        {
+            // Now that we're done with everything, lets see if there are any using statements to fix up
+            // We do this by comparing the original generated C# code, and the changed C# code, and look for a difference
+            // in using statements. We can't use edits for this for two main reasons:
+            //
+            // 1. Using statements in the generated code might come from _Imports.razor, or from this file, and C# will shove them anywhere
+            // 2. The edit might not be clean. eg given:
+            //      using System;
+            //      using System.Text;
+            //    Adding "using System.Linq;" could result in an insert of "Linq;\r\nusing System." on line 2
+            //
+            // So because of the above, we look for a difference in C# using directive nodes directly from the C# syntax tree, and apply them manually
+            // to the Razor document.
+
+            // First grab the old usings. We just convert them all to strings, because we only care about how the statements are represented in code.
+            var oldSyntaxTree = CSharpSyntaxTree.ParseText(csharpText, cancellationToken: cancellationToken);
+            var oldRoot = await oldSyntaxTree.GetRootAsync(cancellationToken);
+            var oldUsings = oldRoot.DescendantNodes(n => n is BaseNamespaceDeclarationSyntax or CompilationUnitSyntax).OfType<UsingDirectiveSyntax>().Select(u => u.ToString().Substring(6));
```

Comment:
Are `UsingDirectiveSyntax` nodes `IEquatable`? Wondering if we can save the SubStringing until the final Workspace edit creation

Reply:
Not in a way that helps us, but I'll write a comparer.

Reply:
Nope, this is way too hard.. not writing a comparer 😛 Will ask and possibly follow up

Reply:
Totally fine! If this only happens on code action resolve now then we can let it be imo 😄